File: construction-platform/python-services/api/error_handler.py
# Enhanced Error Handling
# construction-platform/python-services/api/error_handler.py

from fastapi import HTTPException, Request
from fastapi.responses import JSONResponse
from typing import Dict, Any, Optional
import logging
import traceback
from enum import Enum

logger = logging.getLogger(__name__)

class ErrorType(Enum):
    """Error type enumeration"""
    VALIDATION_ERROR = "validation_error"
    AUTHENTICATION_ERROR = "authentication_error"
    AUTHORIZATION_ERROR = "authorization_error"
    NOT_FOUND_ERROR = "not_found_error"
    RATE_LIMIT_ERROR = "rate_limit_error"
    SERVER_ERROR = "server_error"
    NETWORK_ERROR = "network_error"
    TIMEOUT_ERROR = "timeout_error"
    UNKNOWN_ERROR = "unknown_error"

class ErrorHandler:
    """Enhanced error handler with user-friendly messages"""
    
    ERROR_MESSAGES = {
        ErrorType.VALIDATION_ERROR: "Invalid input data. Please check your request.",
        ErrorType.AUTHENTICATION_ERROR: "Authentication failed. Please check your credentials.",
        ErrorType.AUTHORIZATION_ERROR: "You don't have permission to access this resource.",
        ErrorType.NOT_FOUND_ERROR: "The requested resource was not found.",
        ErrorType.RATE_LIMIT_ERROR: "Too many requests. Please try again later.",
        ErrorType.SERVER_ERROR: "An internal server error occurred. Please try again later.",
        ErrorType.NETWORK_ERROR: "Network error occurred. Please check your connection.",
        ErrorType.TIMEOUT_ERROR: "Request timed out. Please try again.",
        ErrorType.UNKNOWN_ERROR: "An unexpected error occurred. Please contact support."
    }
    
    @staticmethod
    def classify_error(error: Exception) -> ErrorType:
        """Classify error type"""
        error_str = str(error).lower()
        
        if "validation" in error_str or "invalid" in error_str:
            return ErrorType.VALIDATION_ERROR
        elif "authentication" in error_str or "unauthorized" in error_str:
            return ErrorType.AUTHENTICATION_ERROR
        elif "forbidden" in error_str or "permission" in error_str:
            return ErrorType.AUTHORIZATION_ERROR
        elif "not found" in error_str or "404" in error_str:
            return ErrorType.NOT_FOUND_ERROR
        elif "rate limit" in error_str or "429" in error_str:
            return ErrorType.RATE_LIMIT_ERROR
        elif "timeout" in error_str or "timed out" in error_str:
            return ErrorType.TIMEOUT_ERROR
        elif "network" in error_str or "connection" in error_str:
            return ErrorType.NETWORK_ERROR
        else:
            return ErrorType.SERVER_ERROR
    
    @staticmethod
    def create_error_response(
        error: Exception,
        error_type: Optional[ErrorType] = None,
        status_code: int = 500,
        include_details: bool = False
    ) -> JSONResponse:
        """Create user-friendly error response"""
        if error_type is None:
            error_type = ErrorHandler.classify_error(error)
        
        error_message = ErrorHandler.ERROR_MESSAGES.get(
            error_type,
            ErrorHandler.ERROR_MESSAGES[ErrorType.UNKNOWN_ERROR]
        )
        
        response_data = {
            "error": error_type.value,
            "message": error_message,
            "status_code": status_code
        }
        
        # Include details in development mode
        if include_details:
            response_data["details"] = str(error)
            response_data["traceback"] = traceback.format_exc()
        
        # Log error
        logger.error(f"Error: {error_type.value} - {error_message}", exc_info=error)
        
        return JSONResponse(
            status_code=status_code,
            content=response_data
        )
    
    @staticmethod
    async def exception_handler(request: Request, exc: Exception) -> JSONResponse:
        """Global exception handler"""
        # Check if it's an HTTPException
        if isinstance(exc, HTTPException):
            error_type = ErrorHandler.classify_error(exc)
            return ErrorHandler.create_error_response(
                exc,
                error_type,
                exc.status_code,
                include_details=False
            )
        
        # Handle other exceptions
        error_type = ErrorHandler.classify_error(exc)
        status_code = 500 if error_type == ErrorType.SERVER_ERROR else 400
        
        return ErrorHandler.create_error_response(
            exc,
            error_type,
            status_code,
            include_details=False
        )

class RetryStrategy:
    """Retry strategy with exponential backoff"""
    def __init__(self, max_retries: int = 3, base_delay: float = 1.0, max_delay: float = 60.0):
        self.max_retries = max_retries
        self.base_delay = base_delay
        self.max_delay = max_delay
    
    async def retry(self, func, *args, **kwargs):
        """Retry function with exponential backoff"""
        import asyncio
        import random
        
        for attempt in range(self.max_retries):
            try:
                return await func(*args, **kwargs)
            except Exception as e:
                if attempt == self.max_retries - 1:
                    raise
                
                # Calculate delay with exponential backoff and jitter
                delay = min(
                    self.base_delay * (2 ** attempt),
                    self.max_delay
                )
                jitter = random.uniform(0, delay * 0.1)
                delay += jitter
                
                logger.warning(f"Retry attempt {attempt + 1}/{self.max_retries} after {delay:.2f}s")
                await asyncio.sleep(delay)
        
        raise Exception("Max retries exceeded")

File: construction-platform/n8n-workflows/unified/error_handler.js
/**
 * Simplified Error Handler for Construction AI Platform
 * Centralized error handling utility for all workflows
 */

/**
 * Enhanced Error Handler for Construction AI Platform v2.0
 * Supports project context and improved error tracking
 */

class ErrorHandler {
    constructor(context = {}) {
        this.context = context;
        this.executionId = context.executionId || 'unknown';
        this.workflowName = context.workflowName || 'unknown';
        this.projectId = context.projectId || context.project?.id || null;
        this.projectName = context.projectName || context.project?.name || null;
        this.userId = context.userId || context.user?.id || null;
        this.tenantId = context.tenantId || context.tenant?.id || null;
        this.timestamp = new Date().toISOString();
        this.version = '2.0.0';
    }

    /**
     * Handle errors with standardized format
     */
    handleError(error, context = {}) {
        const errorInfo = {
            success: false,
            error: {
                message: this.getErrorMessage(error),
                type: this.getErrorType(error),
                code: this.getErrorCode(error),
                timestamp: this.timestamp,
                executionId: this.executionId,
                workflowName: this.workflowName,
                workflowVersion: this.version,
                // Project context (NEW)
                project: {
                    id: this.projectId,
                    name: this.projectName
                },
                user: {
                    id: this.userId
                },
                tenant: {
                    id: this.tenantId
                },
                context: context
            },
            recovery: {
                suggestions: this.getRecoverySuggestions(error),
                canRetry: this.canRetry(error),
                retryAfter: this.getRetryAfter(error)
            }
        };

        // Enhanced logging with project context
        const logMessage = {
            error: errorInfo.error,
            project: this.projectId ? `Project: ${this.projectId} (${this.projectName})` : 'No project context',
            user: this.userId ? `User: ${this.userId}` : 'Unknown user',
            tenant: this.tenantId ? `Tenant: ${this.tenantId}` : 'Unknown tenant'
        };
        console.error('Error:', logMessage);

        return errorInfo;
    }

    /**
     * Get user-friendly error message
     */
    getErrorMessage(error) {
        if (typeof error === 'string') {
            return error;
        }
        if (error?.message) {
            return error.message;
        }
        if (error?.error?.message) {
            return error.error.message;
        }
        return 'An unexpected error occurred';
    }

    /**
     * Get error type
     */
    getErrorType(error) {
        if (error?.code === 'ENOTFOUND' || error?.code === 'ECONNREFUSED') {
            return 'network_error';
        }
        if (error?.code === 'ETIMEDOUT') {
            return 'timeout_error';
        }
        if (error?.statusCode) {
            if (error.statusCode >= 400 && error.statusCode < 500) {
                return 'client_error';
            }
            if (error.statusCode >= 500) {
                return 'server_error';
            }
        }
        if (error?.message?.includes('not found')) {
            return 'not_found';
        }
        if (error?.message?.includes('unauthorized') || error?.message?.includes('forbidden')) {
            return 'authentication_error';
        }
        if (error?.message?.includes('validation') || error?.message?.includes('invalid')) {
            return 'validation_error';
        }
        return 'unknown_error';
    }

    /**
     * Get error code
     */
    getErrorCode(error) {
        return error?.code || error?.statusCode || 'UNKNOWN_ERROR';
    }

    /**
     * Get recovery suggestions
     */
    getRecoverySuggestions(error) {
        const errorType = this.getErrorType(error);
        const suggestions = {
            network_error: [
                'Check your internet connection',
                'Verify the service is available',
                'Try again in a few moments'
            ],
            timeout_error: [
                'The request took too long',
                'Try again with a smaller file',
                'Check if the service is overloaded'
            ],
            client_error: [
                'Check your input parameters',
                'Verify file format and size',
                'Ensure all required fields are provided'
            ],
            server_error: [
                'Service temporarily unavailable',
                'Try again later',
                'Contact support if problem persists'
            ],
            not_found: [
                'Check if the file/resource exists',
                'Verify the file ID or path',
                'Ensure you have access permissions'
            ],
            authentication_error: [
                'Check your API credentials',
                'Verify authentication tokens',
                'Ensure your account has necessary permissions'
            ],
            validation_error: [
                'Check input format',
                'Verify required fields',
                'Ensure data types are correct'
            ],
            unknown_error: [
                'Try again',
                'Check logs for details',
                'Contact support if problem persists'
            ]
        };

        return suggestions[errorType] || suggestions.unknown_error;
    }

    /**
     * Check if error can be retried
     */
    canRetry(error) {
        const errorType = this.getErrorType(error);
        const retryableErrors = ['network_error', 'timeout_error', 'server_error'];
        return retryableErrors.includes(errorType);
    }

    /**
     * Get retry delay in seconds
     */
    getRetryAfter(error) {
        const errorType = this.getErrorType(error);
        const delays = {
            network_error: 5,
            timeout_error: 10,
            server_error: 30,
            default: 60
        };
        return delays[errorType] || delays.default;
    }

    /**
     * Validate input parameters
     */
    validateInput(input, requiredFields = []) {
        const errors = [];

        // Check required fields
        for (const field of requiredFields) {
            if (!input[field] && input[field] !== 0 && input[field] !== false) {
                errors.push(`Missing required field: ${field}`);
            }
        }

        // Validate file extensions
        if (input.fileExtension) {
            const validExtensions = ['.rvt', '.ifc', '.dwg', '.dgn', '.pdf', '.jpg', '.png', '.docx', '.xlsx', '.csv'];
            const ext = input.fileExtension.toLowerCase();
            if (!validExtensions.includes(ext)) {
                errors.push(`Invalid file extension: ${ext}. Valid extensions: ${validExtensions.join(', ')}`);
            }
        }

        // Validate request type
        if (input.requestType) {
            const validTypes = ['convert', 'validate', 'classify', 'estimate_cost', 'carbon_footprint', 'quantity_takeoff', 'extract_data', 'materials', 'generate_document', 'file_management', 'vendor', 'compliance', 'visualization', 'bim', 'scheduling', '3d_vision'];
            if (!validTypes.includes(input.requestType)) {
                errors.push(`Invalid request type: ${input.requestType}. Valid types: ${validTypes.join(', ')}`);
            }
        }

        if (errors.length > 0) {
            return {
                valid: false,
                errors: errors
            };
        }

        return { valid: true };
    }

    /**
     * Create standardized success response
     */
    createSuccessResponse(data, message = 'Operation completed successfully') {
        return {
            success: true,
            data: data,
            message: message,
            timestamp: this.timestamp,
            executionId: this.executionId
        };
    }

    /**
     * Wrap async function with error handling
     */
    async wrapAsync(fn, context = {}) {
        try {
            const result = await fn();
            return this.createSuccessResponse(result);
        } catch (error) {
            return this.handleError(error, context);
        }
    }
}

// Export for use in n8n workflows
module.exports = ErrorHandler;

